CHECKLIST.md

```diff
@@ -531,11 +531,11 @@ None. No matter how challenging it feels to avoid implementation types in public
 
 [Simplicity](https://theamiableapi.com/2011/09/07/api-design-best-practice-keep-it-simple/) and [Consistency](https://theamiableapi.com/2011/09/14/api-design-best-practice-consistency/). Each API deserves its own package, so that developers can focus on the features they need, without any distracting clutter. The most frequently heard argument for putting several APIs into the same package is their small size. This argument fails to consider that APIs evolve, and what is a small package today can easily become a very large package containing a hodge-podge of unrelated features over time. The java.util package is the best know example.
 
-### 1.1.10.  **Do not** place API and SPI into the same package
+### 1.1.10.  **Avoid** relying on documentation to control evolution of API or SPI
 
-**Rationale:**
-
-[Simplicity](https://theamiableapi.com/2011/09/07/api-design-best-practice-keep-it-simple/), [Consistency](https://theamiableapi.com/2011/09/14/api-design-best-practice-consistency/) and [Evolution](https://theamiableapi.com/2011/10/18/api-design-best-practice-plan-for-evolution/). APIs and SPIs serve different purpose, are used differently, and evolve differently. APIs expose functionality to use. SPIs define functionality to implement and may offer certain facilities to help implement it. A related API and SPI pair may share some of the simpler public support types leading you to believe they belong in the same package. They don’t. Put the shared support types into the API package (making it the self-contained package) and make the SPI package depend on it (making it a bit more complex to use). This is in line with the expectations of most developers, that APIs are easier to use than SPIs.
+SPI implementors can always call their own code, therefore treating it like API. This should be encouraged when they
```

Comment:
Separating API and SPI is a good thing. You can evolve them in a backwards-compatible (constrained by the compiler) way (
see http://wiki.apidesign.org/wiki/ExtendingInterfaces and http://wiki.apidesign.org/wiki/APIvsSPI#Separation_.26_Composition).

Initially, it's more work to define what's SPI and what's API and to keep them separate, but it makes API/SPI more open for evolution in the future.

Reply:
It seems that Practical API Design (PAD) calls API = [ClientAPI](http://wiki.apidesign.org/wiki/ClientAPI) and SPI = [ProviderAPI](http://wiki.apidesign.org/wiki/ProviderAPI).

The difference they highlight is that API can add methods and SPI cannot:
> Slight modifications of the contract are allowed, but compared to ClientAPI (where adding methods is good), the interface for providers is really fixed. No new method additions (adding a new method would put additional, unsatisfiable requirement on already existing implementors), no refactorings. Stability is good. 

I agree with this. They also say:
> do not add new methods to interfaces or abstract classes implemented by some providers.

We will never know what "some providers" implemented or not, so all API interfaces are automatically frozen. And all abstract classes implementable externally do the same.

The only reliable way to allow adding new methods to API is to use abstract classes and forbid external inheritance, ie. [sealed classes](https://dzone.com/articles/effectively-sealed-classes-in-java) if we need polymorphism or final classes if we do not.
The latter is recommended by PAD:
> Exposing only final classes to users of a ClientAPI makes it bullet-proof

Reply:
Note that I heard opinions that SPI allows you to remove methods. It absolutely does not, because it:
* breaks source and biniary compatibility in call-sites (e.g. tests) 
* breaks source compatibility in the implementations (illegal `@Override`)

I agree with PAD, that SPI has more restrictions than API and has no additional freedoms.

Reply:
@dagguh Agree. If you expose an interface for an API, then it becomes SPI. Keeping API and SPI separate means all the API classes are final.

Reply:
@dagguh You're right. I treat "**Do not** place API and SPI into the same package" as a check if you separate both as concepts. Final class vs interface (or abstract class) is the preferred separation method.

Reply:
We should also consider why users prefer to add interfaces for API -> to mock API.

Reply:
[2.3.14](https://github.com/wyrzyk/java-api-design-checklist/blob/82eba514feaa63754b68a3fb05c093f98660249e/CHECKLIST.md#2314-consider-declaring-the-class-as-final) recommends final classes, but it does not justify it with evolution.

[2.3.15](https://github.com/wyrzyk/java-api-design-checklist/blob/82eba514feaa63754b68a3fb05c093f98660249e/CHECKLIST.md#2315-consider-preventing-class-instantiation-by-not-providing-a-public-constructor) recommends hiding the constructor. But it is not enough for a sealed class (inheritance is possible) and it doesn't use evolution rationale.

2.4.3 already defends de facto SPI evolution, but lacks rationale.
2.1.15 and 2.3.18 forbid sealed classes, but without rationale.

I propose to add an explicit rule:
`2.3.23: **Do** use sealed classes for API extensibility`
I'd deprecate rules 2.1.15 and 2.3.18 or add an exception. WDYT?

Reply:
Looks good to me.

Reply:
2.4.3 is a broad issue. Since Java 8 we have default methods. It's not 100% compatible to add a default method, but in practice, we can't always be strict here. We don't have to address all the aspect at once.

2.3.23: **Do** use sealed classes for API extensibility > 2.3.23: **Consider** use sealed classes for API extensibility. WDYT?

Reply:
Sealed classes are not a feature of Java language. It's more trick/hack, I'd not force anyone to use it. But it seems to be a good suggestion.

Reply:
Let's refine 2.4.3 to "Do not add *abstract* methods to a released public Java interface".
When you release API, you choose between `interface` and `final`/sealed `class`. If you design for evolution, you shouldn't choose `interface`, because you'd have to hope 🤞 that the future new method will have a sensible default implementation.
I do not see a way to have both evolution and client-extensibility at the same time.

I found that [2.3.21](https://github.com/wyrzyk/java-api-design-checklist/blob/82eba514feaa63754b68a3fb05c093f98660249e/CHECKLIST.md#2321-do-not-define-public-constructors-for-public-abstract-classes-clients-should-not-extend) already covers sealed classes. It uses package-protected visibility, which is weaker than private, but:
> it is enough to make it clear that what they are doing is unsupported and unsafe

> Sealed classes are not a feature of Java language.

It's an [candidate Java feature](https://openjdk.java.net/jeps/360).

>  I'd not force anyone to use it. But it seems to be a good suggestion.

If you plan to add methods directly, then forcing this seems like the only way. But there are other options, so you are right, this should be a consideration at max. The other option is to [add methods via subtypes](https://github.com/atlassian/infrastructure/commit/4c0cf4012f58bfc66944ac104e0afc2b236cc633#diff-e4d915ff43be479807b343a1d917dced).

Reply:
> We should also consider why developers prefer to add interfaces for API -> to mock API.

Oh man, this is huge. Final and sealed classes are at odds with mockability. Hacks like PowerMockito can work around it. If mocking is hard or impossible, then the provided API should be enough for testing. E.g. DTO types, (which should be `final class`) don't need mocking. If there's mockable behavior in the API, then it seems we have to make it client-extensible and therefore SPI and therefore never add methods to it. Therefore rely on subinterfaces to add methods.

Reply:
There's another alternative, which relies on documentation. AWS API does this:
* [point from the interface to the abstract class](https://github.com/aws/aws-sdk-java/blob/e04c4ac666414a2843385c6d0a52fd57f9c9db52/aws-java-sdk-ec2/src/main/java/com/amazonaws/services/ec2/AmazonEC2.java#L26-L27)
* [implement the interface without real behavior](https://github.com/aws/aws-sdk-java/blob/e04c4ac666414a2843385c6d0a52fd57f9c9db52/aws-java-sdk-ec2/src/main/java/com/amazonaws/services/ec2/AbstractAmazonEC2.java)

The impact on the client is less reliable, because it hopes that clients have read, understood and listened to the advice in the Javadoc.

Reply:
>  If there's mockable behavior in the API, then it seems we have to make it client-extensible and therefore SPI and therefore never add methods to it. 

Interfaces and open classes support mocking out of the box. Testability can be achieved by data classes or some SPIs. But it needs to be designed to be testable. 

Another general approach to make testing easier could be to create sealed classes with package protected constructor and provide another module (test kit) with mocks (the module could be added as a test dependency).